config/routes.rb

```diff
@@ -27,11 +27,15 @@ def matches?(request)
         patch '/:authenticator/:service_id/:account' => 'authenticate#update_config'
 
         get '/:authenticator(/:service_id)/:account/login' => 'authenticate#login'
-        # authn-oidc login & authenticate are currently for future use only
-        #post '/authn-oidc(/:service_id)/:account/login' => 'authenticate#login_oidc'
 
-        # authn-oidc has to be first as it can be ambgiuous with the optional :service_id & :id
+        # authn-oidc has to be first as it can be ambiguous with the optional :service_id in
+        # the common authn request and the fact that authn-oidc doesn't have an 'id' param.
+        # i.e the request 'authn-oidc/:service_id/:account/authenticate' can be interpreted as
+        # ':authenticator/:account/:id/authenticate'
         post '/authn-oidc/:service_id/:account/authenticate' => 'authenticate#authenticate_oidc'
+        # Verify there is a service-id in the request
+        post "/authn-oidc/:account/authenticate" => "authenticate#authn_oidc_missing_service_id"
+
```

Comment:
Is this here for backwards compatibility?  If so, I think we should add a comment saying that.

Reply:
it is not for backwards-compatibility. It's a verification on the feature as there's no use-case of OIDC without service-id

Reply:
i changed the comment to `Verify there is a service-id in the request as it's a requirement for OIDC`

Reply:
> It's a verification on the feature as there's no use-case of OIDC without service-id

In this case, it should simply return 404.  Otherwise aren't we just anticipating a random mistake a user might make?   This kind of special treatment would only make sense if the endpoint without the service_id used to work, and we've now changed it so it doesn't -- hence my question about backward compatibility.  Although that's probably not the most accurate term now that I think about it -- I meant something like "error message to improve UX for users of the now deprecated (non service_id) path".

Reply:
we decided to go with this implementation to have a better UX. we prefer that the user will get the error `Errors::Authentication::AuthnOidc::ServiceIdMissing Service id is required when authenticating with authn-oidc` rather than `ActionController::RoutingError (No route matches [POST] "/authn-oidc/<account>/authenticate")`

Reply:
Yeah I understand the logic, but why do we think a user would make this particular mistake?  Unless there's a prior API version that matched the route without `service_id`, there's no reason to anticipate this.  Think of it this way: by this reasoning, I could start going through every one of our routes, think of possible mistakes a user could make, and add a special route for that mistake with a custom error message -- all in the name of improved UX.

But that's clearly not something we'd want to do.  In addition, this is just not something I've ever seen done.  If you hit a route that doesn't exist, a generic 404 is the standard and correct response.  If we did want to go above and beyond that standard experience, and guess at possible routes a user may have mis-typed or omitted a parameter from, you wouldn't want to do this by adding no-op routes that raise custom errors for each possible mistake.  

Instead, you'd want to add generic code to your `:not_found` handler that looked at the requested non-existent route, compared it to our existing routes, and made suggestions based on that, sort of like a "spell checker" for routing.

Reply:
ok, i'll come back to this PR sometime next week and re-think

Reply:
btw this issue was raised after @telday tried to authenticate with an OIDC authenticator without a service-id. @telday do you want to give your input as a user who encountered this behaviour? what would make it easier for you to understand the issue?

Reply:
@jonahx I disagree here. 
Ellis has raised this so this is why we think a user would make this mistake
if you think of all our authenticators as a whole some of them require service-id and some of them not. It can be confusing at time and it is best that we help the user with this.
if the user came to the actual path of ..../authn-oidc I believe it is because he forgotten something, again as happened to Ellis.
Improving the UX is something we should inspire to.

Reply:
@orenbm with the stated goals here, another alternative would be to leave the original route with the optional `service_id` and handle the empty value in the controller, rather than the additional route.

I agree an additional route for the empty service_id that only returns an error is not a good precedent.

Reply:
> Improving the UX is something we should inspire to.

@InbalZilberman Please re-read my previous post.  I am not objecting to improving the UX.  I am objecting to  implementing that improvement in a way that is inconsistent, and does not align with how routing is intended be used.  I proposed an alternative way to retain the UX improvement without these problems.

Reply:
great! thanks for the clarification

Reply:
changed the behaviour according to this thread and a talk with @sigalsax . thank you all for sharing your thoughts!